lib/saluki-components/src/relays/otlp/mod.rs

```diff
@@ -0,0 +1,194 @@
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl RelayBuilder for OtlpRelayConfiguration {
+    fn output_payload_type(&self) -> PayloadType {
+        PayloadType::Http
+    }
+
+    async fn build(&self, _context: ComponentContext) -> Result<Box<dyn Relay + Send>, GenericError> {
+        Ok(Box::new(OtlpRelay {
+            http_endpoint: self.http_endpoint(),
+            grpc_endpoint: self.grpc_endpoint(),
+            grpc_max_recv_msg_size_bytes: self.grpc_max_recv_msg_size_bytes(),
+            destination_endpoint: self.destination_endpoint().to_string(),
+        }))
+    }
+}
+
+#[async_trait]
+impl Relay for OtlpRelay {
+    async fn run(self: Box<Self>, mut context: RelayContext) -> Result<(), GenericError> {
+        let mut global_shutdown = context.take_shutdown_handle();
+        let mut health = context.take_health_handle();
+        let global_thread_pool = context.topology_context().global_thread_pool().clone();
+        let memory_limiter = context.topology_context().memory_limiter().clone();
+        let dispatcher = context.dispatcher();
+
+        let (payload_tx, mut payload_rx) = mpsc::channel::<OtlpPayload>(1024);
+
+        let handler = RelayHandler::new(payload_tx);
+        let server_builder = OtlpServerBuilder::new(
+            self.http_endpoint.clone(),
+            self.grpc_endpoint.clone(),
+            self.grpc_max_recv_msg_size_bytes,
+        );
+
+        let (http_shutdown, mut http_error) = server_builder
+            .build(handler, memory_limiter.clone(), global_thread_pool.clone())
+            .await?;
+
+        health.mark_ready();
+        debug!(
+            http_endpoint = %self.http_endpoint,
+            grpc_endpoint = %self.grpc_endpoint,
+            destination = %self.destination_endpoint,
+            "OTLP relay started."
+        );
+
+        loop {
+            select! {
+                _ = &mut global_shutdown => {
+                    debug!("Received shutdown signal.");
+                    break
+                },
+                error = &mut http_error => {
+                    if let Some(error) = error {
+                        debug!(%error, "HTTP server error.");
+                    }
+                    break;
+                },
+                Some(otlp_payload) = payload_rx.recv() => {
+                    // Create HTTP payload with destination endpoint
+                    let path = match otlp_payload.signal_type {
+                        OtlpSignalType::Metrics => "/v1/metrics",
+                        OtlpSignalType::Logs => "/v1/logs",
+                        OtlpSignalType::Traces => "/v1/traces",
+                    };
+                    let uri = format!("{}{}", self.destination_endpoint, path);
+                    let mut buffer = ChunkedBytesBuffer::new(4096);
+                    buffer.put_slice(&otlp_payload.data);
+                    let frozen_buffer = buffer.freeze();
```

Comment:
We're going to end up doing a lot of allocating here.

What we should do instead is have `OtlpPayload` hold on to `Bytes` since that's what we're already getting from `OtlpHandler`. With that, we should then add an implementation of `From<Bytes>` to `FrozenChunkedBytesBuffer`, and then that way we can do a nearly zero-overhead conversion here to get things in the right shape for our HTTP request.

As noted in other comments, we can still improve on things even more than this, but this one is basically free to do _right now_.